### PR TITLE
Add more URLayer and URManaged AssociativeCompose instances

### PR DESCRIPTION
**Pull request by sideeffffect**

### CODE REVIEW

File: experimental/shared/src/main/scala/zio/prelude/experimental/EitherCompose.scala [hunk 1]
```diff
@@ -77,6 +77,44 @@ object EitherCompose {
...
+
+  implicit val URLayerEitherCompose: EitherCompose[URLayer] = new EitherCompose[URLayer] {
+
+    type :+:[+l, +r] = Either[l, r]
+
+    def toLeft[A]: URLayer[A, Either[A, Nothing]] = ZLayer.identity[A].map(Left(_))
+
+    def toRight[B]: URLayer[B, Either[Nothing, B]] = ZLayer.identity[B].map(Right(_))
+
+    def fromEither[A, B, C](a2c: => URLayer[A, C])(b2c: => URLayer[B, C]): URLayer[Either[A, B], C] =
```
Comment (sideeffffect) [inline experimental/shared/src/main/scala/zio/prelude/experimental/EitherCompose.scala:89]: Can this be done better? :thinking:

File: experimental/shared/src/main/scala/zio/prelude/experimental/BothCompose.scala [hunk 2]
```diff
@@ -91,6 +91,36 @@ object BothCompose {
       AssociativeCompose.URIOIdentityCompose.compose(bc, ab)
   }
 
+  implicit val URLayerApplicationCompose: BothCompose[URLayer] = new BothCompose[URLayer] {
+
+    type :*:[+f, +s] = Tuple2[f, s]
+
+    def fromFirst[A]: URLayer[(A, Any), A] = ZLayer.identity[(A, Any)].map(_._1)
```
Comment (adamgfraser) [inline experimental/shared/src/main/scala/zio/prelude/experimental/BothCompose.scala:98]:
```suggestion
    def fromFirst[A]: URLayer[(A, Any), A] = ZLayer.first
```

File: experimental/shared/src/main/scala/zio/prelude/experimental/BothCompose.scala [hunk 3]
```diff
@@ -91,6 +91,36 @@ object BothCompose {
...
+
+    def fromSecond[B]: URLayer[(Any, B), B] = ZLayer.identity[(Any, B)].map(_._2)
```
Comment (adamgfraser) [inline experimental/shared/src/main/scala/zio/prelude/experimental/BothCompose.scala:100]:
```suggestion
    def fromSecond[B]: URLayer[(Any, B), B] = ZLayer.second
```

File: experimental/shared/src/main/scala/zio/prelude/experimental/BothCompose.scala [hunk 4]
```diff
@@ -91,6 +91,36 @@ object BothCompose {
...
+
+    def compose[A, B, C](bc: URLayer[B, C], ab: URLayer[A, B]): URLayer[A, C] =
+      AssociativeCompose.URLayerIdentityCompose.compose(bc, ab)
+  }
+
+  implicit val URManagedApplicationCompose: BothCompose[URManaged] = new BothCompose[URManaged] {
+
+    type :*:[+f, +s] = Tuple2[f, s]
+
+    def fromFirst[A]: URManaged[(A, Any), A] = ZManaged.access[(A, Any)](_._1)
```
Comment (adamgfraser) [inline experimental/shared/src/main/scala/zio/prelude/experimental/BothCompose.scala:113]:
```suggestion
    def fromFirst[A]: URManaged[(A, Any), A] =  ZManaged.first
```

File: experimental/shared/src/main/scala/zio/prelude/experimental/BothCompose.scala [hunk 5]
```diff
@@ -91,6 +91,36 @@ object BothCompose {
...
+
+    def fromSecond[B]: URManaged[(Any, B), B] = ZManaged.access[(Any, B)](_._2)
```
Comment (adamgfraser) [inline experimental/shared/src/main/scala/zio/prelude/experimental/BothCompose.scala:115]:
```suggestion
    def fromSecond[B]: URManaged[(Any, B), B] = ZManaged.second
```

File: experimental/shared/src/main/scala/zio/prelude/experimental/EitherCompose.scala [hunk 6]
```diff
@@ -77,6 +77,39 @@ object EitherCompose {
...
+      ZLayer.fromFunctionManyManaged {
+        case Left(a)  => a2c.build.provide(a)
+        case Right(b) => b2c.build.provide(b)
+      }
```
Comment (adamgfraser) [inline experimental/shared/src/main/scala/zio/prelude/experimental/EitherCompose.scala:93]:
```suggestion
    def fromEither[A, B, C](a2c: => URLayer[A, C])(b2c: => URLayer[B, C]): URLayer[Either[A, B], C] = {
      val right: ZLayer[Either[A, B], A, B] =
        ZLayer.fromFunctionManyM[Either[A, B], A, B] {
          case Left(a) => ZIO.fail(a)
          case Right(b) => ZIO.succeed(b)
        }
      val unright: ZLayer[(Either[A, B], Cause[A]), Nothing, A] =
        ZLayer.fromFunctionManyM[(Either[A, B], Cause[A]), Nothing, A]  {
          case (_, cause) => cause.failureOrCause.fold(ZIO.succeed(_), ZIO.halt(_))
        }
      (ZLayer.requires[Either[A, B]] >>> right).fold(unright >>> a2c, b2c)
    }
```

**Review by adamgfraser — APPROVED** [review]
Looks great!

**Comment by sideeffffect** [comment]
we're green :tada:

**Review by adamgfraser — APPROVED** [review]

**Pull request merged**